### ISSUE TITLE
[@types/react] setState type error

### ISSUE THREAD

**Issue by laxip** [DefinitelyTyped/DefinitelyTyped#33697]
I tried using the `@types/react` package and had problems.

See details: https://github.com/facebook/react/issues/15052

```
 public change= (value: number) => {
    this.setState((state) => {
      if(state.condition) {
        return {
          left: value,
        }
      } else {
        return {
          right: value,
        }
      }
    });
  };
```

TS Error:

```
...
        Type '{ left: number; right?: undefined; }' is not assignable to type 'Pick<State, "left" | "right">'.
          Types of property 'right' are incompatible.
            Type 'undefined' is not assignable to type 'number'.
```

@Hotell @johnnyreilly @ericanderson

If you do not mention the authors the issue will be ignored.

**Comment by ericanderson** [comment]
This is a known gap in how the typescript compiler works. When doing conditional logic in a setState callback, you may need to use casting. Given this is not the norm, and that it’s better to be overly strict (and require casts) instead of overly permissive (and allow bugs).

**Comment by TomasHubelbauer** [comment]
I am oblivious to this issue, but at face value, wouldn't `Partial` (which I think used to be used here before `Pick` was introduced in TypeScript) help here?

Is `Pick` providing stronger typing which while having this above drawback is otherwise beneficial so this is accepted so that we can have sounder typing in the general case?

**Comment by benmosher** [comment]
@TomasHubelbauer I came here thinking the same thing but then found this in your linked issue:

https://github.com/DefinitelyTyped/DefinitelyTyped/issues/18365#issuecomment-404048788

tl;dr: the crux is
> a `Partial` is not assignable to a `Pick`

which makes sense upon reflection. `Partial` allows setting `undefined` for any key's value, whereas `Pick` is a strict subset of keys with the same value constraints as the original type.

I found that initializing my update object with lodash's `pick` made the typing straightforward:

```js
const updates = _.pick(this.state, "keyA", "keyB");  // implicitly Pick<State, "keyA" | "keyB">
updates.keyA = newValue;  // can't be undefined unless full state type allows it
updates.keyB = newBValue; // same
this.setState(updates);
```

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
